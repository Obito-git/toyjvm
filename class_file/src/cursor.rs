use crate::ParseError;

pub(crate) struct Cursor<'a> {
    it: std::iter::Copied<std::slice::Iter<'a, u8>>,
}

impl<'a> Cursor<'a> {
    pub fn new(data: &'a [u8]) -> Self {
        Self {
            it: data.iter().copied(),
        }
    }

    fn take<const N: usize>(it: &mut impl Iterator<Item = u8>) -> Option<[u8; N]> {
        let mut buf = [0u8; N];
        for b in &mut buf {
            *b = it.next()?;
        }
        Some(buf)
    }

    pub fn u32(&mut self) -> Result<u32, ParseError> {
        Cursor::take::<4>(&mut self.it)
            .map(u32::from_be_bytes)
            .ok_or(ParseError::UnexpectedEof)
    }

    pub fn u16(&mut self) -> Result<u16, ParseError> {
        Cursor::take::<2>(&mut self.it)
            .map(u16::from_be_bytes)
            .ok_or(ParseError::UnexpectedEof)
    }

    pub fn u8(&mut self) -> Result<u8, ParseError> {
        self.it.next().ok_or(ParseError::UnexpectedEof)
    }

    pub fn bytes(&mut self, n: usize) -> Result<Vec<u8>, ParseError> {
        let mut v = Vec::with_capacity(n);
        for _ in 0..n {
            v.push(self.it.next().ok_or(ParseError::UnexpectedEof)?);
        }
        Ok(v)
    }

    pub fn read_exact(&mut self, buf: &mut [u8]) -> Result<(), ParseError> {
        for byte in buf {
            *byte = self.it.next().ok_or(ParseError::UnexpectedEof)?;
        }
        Ok(())
    }
}
