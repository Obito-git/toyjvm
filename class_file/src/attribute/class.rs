use crate::constant_pool::ConstantPool;
use crate::ClassFileErr;
use common::cursor::ByteCursor;
use std::fmt;
use std::fmt::{Display, Formatter};

pub const ATTR_SOURCE_FILE: &[u8] = b"SourceFile";

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ClassAttribute {
    SourceFile { sourcefile_index: u16 },
    Unknown { name_index: u16, info: Vec<u8> },
}

impl<'a> ClassAttribute {
    pub(crate) fn read(
        pool: &ConstantPool,
        cursor: &mut ByteCursor<'a>,
    ) -> Result<Self, ClassFileErr> {
        let attribute_name_index = cursor.u16()?;
        let attribute_length = cursor.u32()? as usize;

        let utf8 = pool.get_utf8(attribute_name_index)?.as_bytes();
        match utf8 {
            ATTR_SOURCE_FILE => Ok(ClassAttribute::SourceFile {
                sourcefile_index: cursor.u16()?,
            }),
            _ => {
                let mut buf = vec![0u8; attribute_length];
                cursor.read_exact(&mut buf)?;
                Ok(ClassAttribute::Unknown {
                    name_index: attribute_name_index,
                    info: buf,
                })
            }
        }
    }
}

impl Display for ClassAttribute {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            ClassAttribute::SourceFile { sourcefile_index } => {
                write!(f, "SourceFile(sourcefile_index: {})", sourcefile_index)
            }
            ClassAttribute::Unknown { name_index, info } => write!(
                f,
                "Unsupported(name_index: {}, data: {} bytes)",
                name_index,
                info.len()
            ),
        }
    }
}
