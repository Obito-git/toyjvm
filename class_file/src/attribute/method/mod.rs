use crate::ClassFileErr;
use crate::attribute::method::code::CodeAttributeInfo;
use crate::attribute::{AttributeType, SharedAttribute};
use crate::constant::pool::ConstantPool;
use common::descriptor::MethodDescriptor;
use common::instruction::Instruction;
use common::utils::cursor::ByteCursor;
#[cfg(test)]
use serde::Serialize;

pub mod code;

/// https://docs.oracle.com/javase/specs/jvms/se24/html/jvms-4.html#jvms-4.7.3
#[cfg_attr(test, derive(Serialize))]
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ExceptionTableEntry {
    start_pc: u16,
    end_pc: u16,
    handler_pc: u16,
    catch_type: u16,
}

/// https://docs.oracle.com/javase/specs/jvms/se24/html/jvms-4.html#jvms-4.7.3
#[cfg_attr(test, derive(Serialize))]
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CodeAttribute {
    pub max_stack: u16,
    pub max_locals: u16,
    pub code: Vec<u8>,
    pub exception_table: Vec<ExceptionTableEntry>,
    pub attributes: Vec<CodeAttributeInfo>,
}

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MethodParameterEntry {
    pub name_index: u16,
    pub access_flags: u16,
}

/// https://docs.oracle.com/javase/specs/jvms/se24/html/jvms-4.html#jvms-4.7
#[cfg_attr(test, derive(Serialize))]
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum MethodAttribute {
    Shared(SharedAttribute),
    Code(CodeAttribute),
    Exceptions(Vec<u16>),
    RuntimeVisibleParameterAnnotations,
    RuntimeInvisibleParameterAnnotations,
    AnnotationsDefault,
    MethodParameters(Vec<MethodParameterEntry>),
}

impl MethodParameterEntry {
    pub fn new(name_index: u16, access_flags: u16) -> Self {
        Self {
            name_index,
            access_flags,
        }
    }
}

impl<'a> MethodAttribute {
    pub(crate) fn read(
        pool: &ConstantPool,
        cursor: &mut ByteCursor<'a>,
    ) -> Result<Self, ClassFileErr> {
        let attribute_name_index = cursor.u16()?;
        let _attribute_length = cursor.u32()? as usize;

        let attribute_type = AttributeType::try_from(pool.get_utf8(&attribute_name_index)?)?;
        match attribute_type {
            AttributeType::Code => Ok(MethodAttribute::Code(CodeAttribute::read(pool, cursor)?)),
            AttributeType::RuntimeVisibleAnnotations
            | AttributeType::Synthetic
            | AttributeType::Deprecated
            | AttributeType::Signature => Ok(MethodAttribute::Shared(SharedAttribute::read(
                attribute_type,
                cursor,
            )?)),
            AttributeType::MethodParameters => {
                let parameters_count = cursor.u8()? as usize;
                let mut parameters = Vec::with_capacity(parameters_count);
                for _ in 0..parameters_count {
                    parameters.push(MethodParameterEntry::new(cursor.u16()?, cursor.u16()?));
                }
                Ok(MethodAttribute::MethodParameters(parameters))
            }
            AttributeType::Exceptions => {
                let number_of_exceptions = cursor.u16()?;
                let mut exception_index_table = Vec::with_capacity(number_of_exceptions as usize);
                for _ in 0..number_of_exceptions {
                    exception_index_table.push(cursor.u16()?);
                }
                Ok(MethodAttribute::Exceptions(exception_index_table))
            }
            other => unimplemented!("Method attribute {:?} not implemented", other),
        }
    }

    #[cfg(feature = "pretty_print")]
    pub(crate) fn fmt_pretty(
        &self,
        ind: &mut common::utils::indent_write::Indented<'_>,
        cp: &ConstantPool,
        descriptor: &MethodDescriptor,
        this: &u16,
    ) -> std::fmt::Result {
        use common::pretty_try;
        use std::fmt::Write as _;
        match self {
            MethodAttribute::Shared(shared) => shared.fmt_pretty(ind, cp)?,
            MethodAttribute::Code(code) => code.fmt_pretty(ind, cp, descriptor, this)?,
            MethodAttribute::Exceptions(exc) => {
                writeln!(ind, "Exceptions:")?;
                ind.with_indent(|ind| {
                    writeln!(
                        ind,
                        "throws {}",
                        pretty_try!(
                            ind,
                            exc.iter()
                                .map(|index| cp.get_pretty_class_name(index))
                                .collect::<Result<Vec<_>, _>>()
                        )
                        .join(", ")
                    )?;
                    Ok(())
                })?
            }
            MethodAttribute::RuntimeVisibleParameterAnnotations => unimplemented!(),
            MethodAttribute::RuntimeInvisibleParameterAnnotations => unimplemented!(),
            MethodAttribute::AnnotationsDefault => unimplemented!(),
            MethodAttribute::MethodParameters(params) => {
                const W_NAME: usize = 32;
                writeln!(ind, "MethodParameters:")?;
                ind.with_indent(|ind| {
                    writeln!(ind, "{:>W_NAME$} Flags", "Name")?;
                    for param in params {
                        let name = if param.name_index == 0 {
                            "<no name>".to_string()
                        } else {
                            pretty_try!(ind, cp.get_utf8(&param.name_index)).to_string()
                        };
                        writeln!(ind, "{:>W_NAME$} 0x{:04x}", name, param.access_flags,)?;
                    }
                    Ok(())
                })?;
            }
        }

        Ok(())
    }
}

impl<'a> CodeAttribute {
    pub(crate) fn read(
        pool: &ConstantPool,
        cursor: &mut ByteCursor<'a>,
    ) -> Result<Self, ClassFileErr> {
        let max_stack = cursor.u16()?;
        let max_locals = cursor.u16()?;
        let code_length = cursor.u32()? as usize;

        let mut code = vec![0u8; code_length];
        cursor.read_exact(&mut code)?;

        let exception_table_length = cursor.u16()? as usize;
        let mut exception_table = Vec::with_capacity(exception_table_length);
        for _ in 0..exception_table_length {
            exception_table.push(ExceptionTableEntry {
                start_pc: cursor.u16()?,
                end_pc: cursor.u16()?,
                handler_pc: cursor.u16()?,
                catch_type: cursor.u16()?,
            });
        }

        let attributes_count = cursor.u16()? as usize;
        let mut attributes = Vec::with_capacity(attributes_count);
        for _ in 0..attributes_count {
            attributes.push(CodeAttributeInfo::read(pool, cursor)?);
        }

        Ok(Self {
            max_stack,
            max_locals,
            code,
            exception_table,
            attributes,
        })
    }

    #[cfg(feature = "pretty_print")]
    pub(crate) fn fmt_pretty(
        &self,
        ind: &mut common::utils::indent_write::Indented<'_>,
        cp: &ConstantPool,
        method_descriptor: &MethodDescriptor,
        this: &u16,
    ) -> std::fmt::Result {
        use crate::print::get_pretty_instruction;
        use common::pretty_try;
        use std::fmt::Write as _;

        writeln!(ind, "Code: ")?;
        ind.with_indent(|ind| {
            writeln!(
                ind,
                "stack={}, locals={}, args_size={}",
                self.max_stack,
                self.max_locals,
                method_descriptor.params.len() + 1 // +1 for 'this'
            )?;
            let instructions = pretty_try!(ind, Instruction::new_instruction_set(&self.code));
            let mut byte_pos = 0;
            for instruction in instructions {
                writeln!(
                    ind,
                    "{byte_pos:4}: {:<24}",
                    pretty_try!(
                        ind,
                        get_pretty_instruction(&instruction, cp, byte_pos as i32, this)
                    )
                )?;
                byte_pos += instruction.byte_size();
            }
            if !self.exception_table.is_empty() {
                const W_START: usize = 6;
                const W_LENGTH: usize = 8;
                const W_SLOT: usize = 5;
                writeln!(ind, "Exception table:")?;
                ind.with_indent(|ind| {
                    writeln!(
                        ind,
                        "{:>W_START$} {:>W_LENGTH$} {:>W_SLOT$} type",
                        "from", "to", "target"
                    )?;
                    for entry in &self.exception_table {
                        let catch_type = if entry.catch_type == 0 {
                            "any"
                        } else {
                            pretty_try!(ind, cp.get_class_name(&entry.catch_type))
                        };
                        writeln!(
                            ind,
                            "{:>W_START$} {:>W_LENGTH$} {:>W_SLOT$}  {}{}",
                            entry.start_pc,
                            entry.end_pc,
                            entry.handler_pc,
                            if catch_type != "any" { "Class " } else { "" },
                            catch_type
                        )?;
                    }
                    Ok(())
                })?;
            }
            for attr in &self.attributes {
                attr.fmt_pretty(ind, cp, this)?;
            }
            Ok(())
        })?;

        Ok(())
    }
}
