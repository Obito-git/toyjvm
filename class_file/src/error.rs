use crate::attribute::AttributeType;
use crate::constant::ConstantTag;
use common::utils::cursor::CursorError;
#[cfg(test)]
use serde::Serialize;
use thiserror::Error;

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug, Clone, PartialEq, Eq, Error)]
pub enum ClassFileErr {
    #[error(transparent)]
    Cursor(#[from] CursorError),
    #[error("")]
    WrongMagic,
    #[error("Expected end of file but found trailing bytes.")]
    TrailingBytes,
    #[error("")]
    UnknownTag(u8),
    #[error("Expected type `{1}` with index `{0}` but found `{2}`")]
    /// First u16 is index, second is expected type, third is actual type
    TypeError(u16, ConstantTag, ConstantTag),
    #[error("Constant with index `{0}` isn't found in constant constant.")]
    ConstantNotFound(u16),
    #[error("Unknown stack frame type {0}.")]
    UnknownStackFrameType(u8),
    #[error("Unknown attribute `{0}.")]
    UnknownAttribute(String),
    #[error("Can't build shared attribute, the `{0}` attribute isn't shared.")]
    AttributeIsNotShared(AttributeType),
    #[error("Invalid method handle kind {0}.")]
    InvalidMethodHandleKind(u8),
}
