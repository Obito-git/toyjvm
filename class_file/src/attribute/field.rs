use crate::attribute::{AttributeType, SharedAttribute};
use crate::constant::pool::ConstantPool;
use crate::error::ClassFormatErr;
use common::utils::cursor::ByteCursor;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum FieldAttribute {
    Shared(SharedAttribute),
    ConstantValue(u16),
}

impl<'a> FieldAttribute {
    pub(crate) fn read(
        pool: &ConstantPool,
        cursor: &mut ByteCursor<'a>,
    ) -> Result<Self, ClassFormatErr> {
        let attribute_name_index = cursor.u16()?;
        let _attribute_length = cursor.u32()? as usize;

        let attribute_type = AttributeType::try_from(pool.get_utf8(&attribute_name_index)?)?;
        match attribute_type {
            AttributeType::ConstantValue => Ok(FieldAttribute::ConstantValue(cursor.u16()?)),
            AttributeType::RuntimeVisibleAnnotations
            | AttributeType::Synthetic
            | AttributeType::Deprecated
            | AttributeType::Signature => Ok(FieldAttribute::Shared(SharedAttribute::read(
                attribute_type,
                cursor,
            )?)),
            _ => unimplemented!(),
        }
    }

    #[cfg(feature = "pretty_print")]
    pub(crate) fn fmt_pretty(
        &self,
        ind: &mut common::utils::indent_write::Indented<'_>,
        cp: &ConstantPool,
    ) -> std::fmt::Result {
        use common::pretty_try;
        use std::fmt::Write as _;

        match self {
            FieldAttribute::Shared(shared) => shared.fmt_pretty(ind, cp)?,
            FieldAttribute::ConstantValue(val) => {
                let constant = pretty_try!(ind, cp.get_raw(val));
                writeln!(
                    ind,
                    "ConstantValue: {}",
                    pretty_try!(ind, constant.get_pretty_value(cp, &0))
                )?;
            }
        }

        Ok(())
    }
}
