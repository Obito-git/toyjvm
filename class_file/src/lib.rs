use crate::attribute::class::ClassAttribute;
use crate::constant::pool::ConstantPool;
use crate::error::ClassFileErr;
use common::utils::cursor::ByteCursor;
use constant::ConstantInfo;
use field::FieldInfo;
use method::MethodInfo;
#[cfg(test)]
use serde::Serialize;

pub mod attribute;
pub mod constant;
pub mod error;
pub mod field;
pub mod method;
#[cfg(feature = "pretty_print")]
pub mod print;
// TODO: review all access levels in the crate (methods, fields, modules, structs, etc.)
// TODO: align enums that end with "Info"/"Ref" and "Type"/"Kind" suffixes

/// https://docs.oracle.com/javase/specs/jvms/se24/html/jvms-4.html
/// A rust representation of a Java .class file. All structures in the crates have public only public
/// fields for easier access, because anyway it will be remapped to runtime structures.
///
/// All print related code is behind the `pretty_print` feature flag.
#[cfg_attr(test, derive(Serialize))]
#[derive(Debug)]
pub struct ClassFile {
    pub minor_version: u16,
    pub major_version: u16,
    pub cp: ConstantPool,
    pub access_flags: u16,
    pub this_class: u16,
    pub super_class: u16,
    pub interfaces: Vec<u16>,
    pub fields: Vec<FieldInfo>,
    pub methods: Vec<MethodInfo>,
    pub attributes: Vec<ClassAttribute>,
}

impl ClassFile {
    const MAGIC: u32 = 0xCAFEBABE;
    fn validate_magic(val: u32) -> Result<(), ClassFileErr> {
        (val == ClassFile::MAGIC)
            .then_some(())
            .ok_or(ClassFileErr::WrongMagic)
    }
}

impl TryFrom<Vec<u8>> for ClassFile {
    type Error = ClassFileErr;

    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
        let mut cursor = ByteCursor::new(&value);
        let magic = cursor.u32()?;
        ClassFile::validate_magic(magic)?;
        let minor_version = cursor.u16()?;
        let major_version = cursor.u16()?;
        let constant_pool_count = cursor.u16()?;
        let mut constant_pool = Vec::with_capacity((constant_pool_count + 1) as usize);
        constant_pool.push(ConstantInfo::Unused);
        let mut i = 1;
        while i < constant_pool_count {
            let constant = ConstantInfo::read(&mut cursor)?;
            constant_pool.push(constant.clone());
            match constant {
                // described in JVM spec that Long and Double take two entries in the constant pool
                // https://docs.oracle.com/javase/specs/jvms/se24/html/jvms-4.html#jvms-4.4.5
                ConstantInfo::Long(_) | ConstantInfo::Double(_) => {
                    constant_pool.push(ConstantInfo::Unused);
                    i += 2;
                }
                _ => {
                    i += 1;
                }
            }
        }
        let constant_pool = ConstantPool {
            inner: constant_pool,
        };

        let access_flags = cursor.u16()?;
        let this_class = cursor.u16()?;
        let super_class = cursor.u16()?;
        let interfaces_count = cursor.u16()?;
        let mut interfaces = Vec::with_capacity(interfaces_count as usize);
        for _ in 0..interfaces_count {
            interfaces.push(cursor.u16()?);
        }
        let fields_count = cursor.u16()?;
        let mut fields = Vec::with_capacity(fields_count as usize);
        for _ in 0..fields_count {
            fields.push(FieldInfo::read(&constant_pool, &mut cursor)?);
        }
        let methods_count = cursor.u16()?;
        let mut methods = Vec::with_capacity(methods_count as usize);
        for _ in 0..methods_count {
            methods.push(MethodInfo::read(&constant_pool, &mut cursor)?);
        }
        let attributes_count = cursor.u16()?;
        let mut attributes = Vec::with_capacity(attributes_count as usize);
        for _ in 0..attributes_count {
            attributes.push(ClassAttribute::read(&constant_pool, &mut cursor)?);
        }

        if cursor.u8().is_ok() {
            Err(ClassFileErr::TrailingBytes)
        } else {
            Ok(Self {
                minor_version,
                major_version,
                cp: constant_pool,
                access_flags,
                this_class,
                super_class,
                interfaces,
                fields,
                methods,
                attributes,
            })
        }
    }
}

#[cfg(feature = "pretty_print")]
impl std::fmt::Display for ClassFile {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use crate::print::{get_class_javap_like_list, get_class_pretty_java_like_prefix};
        use common::utils::indent_write::Indented;
        use common::{pretty_class_name_try, pretty_try};
        use std::fmt::Write as _;

        const CONSTANT_KIND_WIDTH: usize = 18;
        let mut ind = Indented::new(f);
        writeln!(
            ind,
            "{} {}",
            get_class_pretty_java_like_prefix(self.access_flags),
            pretty_class_name_try!(ind, self.cp.get_class_name(&self.this_class))
        )?;
        ind.with_indent(|ind| {
            writeln!(ind, "minor version: {}", self.minor_version)?;
            writeln!(ind, "major version: {}", self.major_version)?;
            writeln!(
                ind,
                "flags: (0x{:04X}) {}",
                self.access_flags,
                get_class_javap_like_list(self.access_flags)
            )?;
            writeln!(
                ind,
                "this_class: {:<24}//{}",
                format!("#{}", self.this_class),
                pretty_try!(ind, self.cp.get_class_name(&self.this_class))
            )?;
            writeln!(ind, "super_class: #{}", self.super_class)?;
            writeln!(
                ind,
                "interfaces: {}, fields: {}, methods: {}, attributes: {}",
                self.interfaces.len(),
                self.fields.len(),
                self.methods.len(),
                self.attributes.len()
            )?;
            Ok(())
        })?;
        writeln!(ind, "Constant pool:")?;
        ind.with_indent(|ind| {
            let counter_width = self
                .cp
                .inner
                .len()
                .checked_ilog10()
                .map_or(0, |d| d as usize)
                + 2;
            for (i, c) in self.cp.inner.iter().enumerate() {
                if matches!(c, ConstantInfo::Unused) {
                    continue;
                }
                let tag = format_args!("{:<kw$}", c.get_tag(), kw = CONSTANT_KIND_WIDTH);
                write!(ind, "{:>w$} = {} ", format!("#{i}"), tag, w = counter_width)?;
                c.fmt_pretty(ind, &self.cp)?;
            }
            Ok(())
        })?;
        writeln!(ind, "{{")?;
        ind.with_indent(|ind| {
            for (i, field) in self.fields.iter().enumerate() {
                field.fmt_pretty(ind, &self.cp)?;
                if i + 1 < self.methods.len() {
                    writeln!(ind)?;
                }
            }
            Ok(())
        })?;
        writeln!(ind, "}}")?;
        writeln!(ind, "{{")?;
        ind.with_indent(|ind| {
            for (i, method) in self.methods.iter().enumerate() {
                method.fmt_pretty(ind, &self.cp, &self.this_class)?;
                if i + 1 < self.methods.len() {
                    writeln!(ind)?;
                }
            }
            Ok(())
        })?;
        writeln!(ind, "}}")?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::with_settings;
    use rstest::*;
    use std::fs;
    use std::io::BufRead;
    use std::path::{Path, PathBuf};

    const CLASS_SNAPSHOT_PATH: &str = "../snapshots/class_file";
    const DISPLAY_SNAPSHOT_PATH: &str = "../snapshots/display";

    fn to_snapshot_name(path: &Path) -> String {
        let mut iter = path.iter().map(|s| s.to_string_lossy().to_string());
        for seg in iter.by_ref() {
            if seg == "test-classes" {
                break;
            }
        }
        let tail: Vec<String> = iter.collect();
        tail.join("-")
    }

    #[rstest]
    #[trace]
    // Requires `testdata/compile-fixtures.py` to be executed to generate the .class files
    fn parse_class_file(
        #[base_dir = "../target/test-classes"]
        #[files("**/*.class")]
        path: PathBuf,
    ) {
        // Given
        let bytes = fs::read(&path).unwrap_or_else(|_| panic!("Can't read file {:?}", path));

        // When
        let class_file = ClassFile::try_from(bytes).unwrap();

        // Then
        /*
        with_settings!(
            {
                snapshot_path => CLASS_SNAPSHOT_PATH,
                prepend_module_to_snapshot => false,
            },
            {
                insta::assert_yaml_snapshot!(to_snapshot_name(&path), class_file);
            }
        );
        */
    }

    #[rstest]
    #[trace]
    // Requires `testdata/compile-fixtures.sh` to be run to generate the .class files
    fn test_display(
        #[base_dir = "../target/test-classes"]
        #[files("**/*.class")]
        path: PathBuf,
    ) {
        // Given
        let bytes = fs::read(&path).unwrap_or_else(|_| panic!("Can't read file {:?}", path));

        // When
        let display = format!("{}", ClassFile::try_from(bytes).unwrap());

        // Then
        /*
        with_settings!(
            {
                snapshot_path => DISPLAY_SNAPSHOT_PATH,
                prepend_module_to_snapshot => false,
            },
            {
                insta::assert_snapshot!(to_snapshot_name(&path), display);
            }
        );
         */
    }

    #[rstest]
    // Requires `testdata/compile-fixtures.sh` to be run to generate the .class files
    fn compare_with_javap(
        #[base_dir = "../target/test-classes"]
        #[files("**/*.class")]
        path: PathBuf,
    ) {
        // Given
        let bytes = fs::read(&path).unwrap_or_else(|_| panic!("Can't read file {:?}", path));
        let my_display = format!("{}", ClassFile::try_from(bytes).unwrap())
            .as_bytes()
            .lines()
            .collect::<Result<Vec<_>, _>>()
            .unwrap();
        let javap = std::process::Command::new("javap")
            .arg("-v")
            .arg("-p")
            .arg(&path)
            .output()
            .unwrap_or_else(|e| panic!("Can't run javap on file {:?}:{}\n", path, e))
            .stdout
            .lines()
            .collect::<Result<Vec<_>, _>>()
            .unwrap();
        let javap_display = javap[4..javap.len() - 1].to_vec();

        // When && Then
        for (i, (my, javap)) in my_display.iter().zip(javap_display.iter()).enumerate() {
            let my_line_normalized: String = my.chars().filter(|c| !c.is_whitespace()).collect();
            let javap_line_normalized: String =
                javap.chars().filter(|c| !c.is_whitespace()).collect();
            assert_eq!(
                my_line_normalized,
                javap_line_normalized,
                "Mismatch at line {} of file {:?}",
                i + 1,
                path,
            );
        }
    }
}
