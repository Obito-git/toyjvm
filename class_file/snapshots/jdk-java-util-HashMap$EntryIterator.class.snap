---
source: class_file/src/lib.rs
expression: display
---
final class java.util.HashMap$EntryIterator extends java.util.HashMap<K, V>.HashIterator implements java.util.Iterator<java.util.Map$Entry<K, V>>
  minor version: 0
  major version: 68
  flags: (0x0030) ACC_FINAL, ACC_SUPER
  this_class: #8                       //java/util/HashMap$EntryIterator
  super_class: #2                      //java/util/HashMap$HashIterator
  interfaces: 1, fields: 0, methods: 3, attributes: 4
Constant pool:
   #1 = Methodref          #2.#3            // java/util/HashMap$HashIterator."<init>":(Ljava/util/HashMap;)V
   #2 = Class              #4               // java/util/HashMap$HashIterator
   #3 = NameAndType        #5:#6            // "<init>":(Ljava/util/HashMap;)V
   #4 = Utf8               java/util/HashMap$HashIterator
   #5 = Utf8               <init>
   #6 = Utf8               (Ljava/util/HashMap;)V
   #7 = Methodref          #8.#9            // java/util/HashMap$EntryIterator.nextNode:()Ljava/util/HashMap$Node;
   #8 = Class              #10              // java/util/HashMap$EntryIterator
   #9 = NameAndType        #11:#12          // nextNode:()Ljava/util/HashMap$Node;
  #10 = Utf8               java/util/HashMap$EntryIterator
  #11 = Utf8               nextNode
  #12 = Utf8               ()Ljava/util/HashMap$Node;
  #13 = Methodref          #8.#14           // java/util/HashMap$EntryIterator.next:()Ljava/util/Map$Entry;
  #14 = NameAndType        #15:#16          // next:()Ljava/util/Map$Entry;
  #15 = Utf8               next
  #16 = Utf8               ()Ljava/util/Map$Entry;
  #17 = Class              #18              // java/util/Iterator
  #18 = Utf8               java/util/Iterator
  #19 = Utf8               Code
  #20 = Utf8               LineNumberTable
  #21 = Utf8               LocalVariableTable
  #22 = Utf8               this
  #23 = Utf8               Ljava/util/HashMap$EntryIterator;
  #24 = Utf8               this$0
  #25 = Utf8               Ljava/util/HashMap;
  #26 = Utf8               LocalVariableTypeTable
  #27 = Utf8               Ljava/util/HashMap<TK;TV;>.EntryIterator;
  #28 = Utf8               MethodParameters
  #29 = Utf8               Signature
  #30 = Utf8               ()Ljava/util/Map$Entry<TK;TV;>;
  #31 = Utf8               ()Ljava/lang/Object;
  #32 = Utf8               Ljava/util/HashMap<TK;TV;>.HashIterator;Ljava/util/Iterator<Ljava/util/Map$Entry<TK;TV;>;>;
  #33 = Utf8               SourceFile
  #34 = Utf8               HashMap.java
  #35 = Utf8               NestHost
  #36 = Class              #37              // java/util/HashMap
  #37 = Utf8               java/util/HashMap
  #38 = Utf8               InnerClasses
  #39 = Utf8               HashIterator
  #40 = Utf8               EntryIterator
  #41 = Class              #42              // java/util/HashMap$Node
  #42 = Utf8               java/util/HashMap$Node
  #43 = Utf8               Node
  #44 = Class              #45              // java/util/Map$Entry
  #45 = Utf8               java/util/Map$Entry
  #46 = Class              #47              // java/util/Map
  #47 = Utf8               java/util/Map
  #48 = Utf8               Entry
{
  java.util.HashMap$EntryIterator(java.util.HashMap);
    descriptor: (Ljava/util/HashMap;)V
    flags: (0x0000) 
    Code: 
      stack=2, locals=2, args_size=2
         0: aload_0                 
         1: aload_1                 
         2: invokespecial #1                  // Method java/util/HashMap$HashIterator."<init>":(Ljava/util/HashMap;)V
         5: return                  
      LineNumberTable:
        line 1637: 0
      LocalVariableTable:
       Start   Length  Slot  Name   Signature
           0        6     0  this   Ljava/util/HashMap$EntryIterator;
           0        6     1  this$0 Ljava/util/HashMap;
      LocalVariableTypeTable:
         Start   Length  Slot  Name   Signature
             0        6     0  this   Ljava/util/HashMap<TK;TV;>.EntryIterator;
    MethodParameters:
      Name                             Flags
      <no name>                        final mandated 
  
  public final java.util.Map$Entry<K, V> next();
    descriptor: ()Ljava/util/Map$Entry;
    flags: (0x0011) ACC_PUBLIC, ACC_FINAL
    Code: 
      stack=1, locals=1, args_size=1
         0: aload_0                 
         1: invokevirtual #7                  // Method nextNode:()Ljava/util/HashMap$Node;
         4: areturn                 
      LineNumberTable:
        line 1639: 0
      LocalVariableTable:
       Start   Length  Slot  Name   Signature
           0        5     0  this   Ljava/util/HashMap$EntryIterator;
      LocalVariableTypeTable:
         Start   Length  Slot  Name   Signature
             0        5     0  this   Ljava/util/HashMap<TK;TV;>.EntryIterator;
    Signature: #30                         // ()Ljava/util/Map$Entry<TK;TV;>;
  
  public java.lang.Object next();
    descriptor: ()Ljava/lang/Object;
    flags: (0x1041) ACC_PUBLIC, ACC_BRIDGE, ACC_SYNTHETIC
    Code: 
      stack=1, locals=1, args_size=1
         0: aload_0                 
         1: invokevirtual #13                 // Method next:()Ljava/util/Map$Entry;
         4: areturn                 
      LineNumberTable:
        line 1637: 0
      LocalVariableTable:
       Start   Length  Slot  Name   Signature
           0        5     0  this   Ljava/util/HashMap$EntryIterator;
      LocalVariableTypeTable:
         Start   Length  Slot  Name   Signature
             0        5     0  this   Ljava/util/HashMap<TK;TV;>.EntryIterator;
}
Signature: #32                         // Ljava/util/HashMap<TK;TV;>.HashIterator;Ljava/util/Iterator<Ljava/util/Map$Entry<TK;TV;>;>;
SourceFile: "HashMap.java"
NestHost: class java/util/HashMap
InnerClasses:
  abstract #39= #2 of #36;                    // HashIterator=class java/util/HashMap$HashIterator of class java/util/HashMap
  final #40= #8 of #36;                       // EntryIterator=class java/util/HashMap$EntryIterator of class java/util/HashMap
  static #43= #41 of #36;                     // Node=class java/util/HashMap$Node of class java/util/HashMap
  public static #48= #44 of #46;              // Entry=class java/util/Map$Entry of class java/util/Map
